### PR TITLE
Bug 1822127 - Fix registry proxy secret environment injection

### PR DESCRIPTION
**Pull request by dymurray**

### CODE REVIEW

File: pkg/apis/migration/v1alpha1/migplan_types.go [hunk 1]
```diff
@@ -371,36 +371,51 @@ func (r *MigPlan) BuildRegistryDC(storage *MigStorage, name, dirName string) (*a
...
 	envVars := []kapi.EnvVar{}
-	if found, s := Settings.HasProxyVar(settings.HttpProxy); found {
+	secret, err := GetSecret(
+		client,
+		&kapi.ObjectReference{
+			Namespace: VeleroNamespace,
+			Name:      ProxySecret,
+		},
+	)
+	if err != nil {
```
Comment (jortel) [inline pkg/apis/migration/v1alpha1/migplan_types.go:389]: The `GetSecret()` will return a `nil` secret when `NotFound`.  Any other error should be returned.

File: pkg/apis/migration/v1alpha1/migplan_types.go [hunk 2]
```diff
@@ -371,36 +371,51 @@ func (r *MigPlan) BuildRegistryDC(storage *MigStorage, name, dirName string) (*a
...
+// Build up proxy env vars for registry DC
+// Returns empty array if secret isn't found or no configuration exists
+func (r *MigPlan) BuildRegistryProxyEnvVars(client k8sclient.Client) []kapi.EnvVar {
 	envVars := []kapi.EnvVar{}
-	if found, s := Settings.HasProxyVar(settings.HttpProxy); found {
+	secret, err := GetSecret(
+		client,
+		&kapi.ObjectReference{
+			Namespace: VeleroNamespace,
+			Name:      ProxySecret,
```
Comment (jortel) [inline pkg/apis/migration/v1alpha1/migplan_types.go:386]: What creates the proxy secret?
I'm concerned about the _special_ name.  Usually we use labels.

Comment (dymurray) [reply pkg/apis/migration/v1alpha1/migplan_types.go:386]: @eriknelson Let's move to attaching a label to this secret. it's fine to keep it `migration-proxy` but lets also add a label.

File: pkg/apis/migration/v1alpha1/migplan_types.go [hunk 3]
```diff
@@ -371,36 +372,68 @@ func (r *MigPlan) BuildRegistryDC(storage *MigStorage, name, dirName string) (*a
...
 	}
-	if found, s := Settings.HasProxyVar(settings.NoProxy); found {
-		envVars = append(envVars, kapi.EnvVar{
-			Name:  settings.NoProxy,
-			Value: s,
-		})
+	if len(list.Items) > 1 {
+		return nil, errors.New("found multiple proxy config secrets")
+	}
+	secret, err := GetSecret(
```
Comment (jortel) [inline pkg/apis/migration/v1alpha1/migplan_types.go:406]: Why fetch the secret when it's already in list.Items[0]?

File: pkg/apis/migration/v1alpha1/migplan_types.go [hunk 4]
```diff
@@ -371,36 +372,68 @@ func (r *MigPlan) BuildRegistryDC(storage *MigStorage, name, dirName string) (*a
...
+	// This means no proxy is configured
+	if len(list.Items) == 0 {
+		return nil, nil
 	}
-	if found, s := Settings.HasProxyVar(settings.NoProxy); found {
-		envVars = append(envVars, kapi.EnvVar{
-			Name:  settings.NoProxy,
-			Value: s,
-		})
+	if len(list.Items) > 1 {
```
Comment (jortel) [inline pkg/apis/migration/v1alpha1/migplan_types.go:403]: Thinking a validation & condition (in validations) would be better than returning an error here.  The logged trace and (failed) reconcile and reQ will be messy.

File: pkg/apis/migration/v1alpha1/migplan_types.go [hunk 5]
```diff
@@ -371,36 +372,68 @@ func (r *MigPlan) BuildRegistryDC(storage *MigStorage, name, dirName string) (*a
...
 // Update a Registry DeploymentConfig as desired for the specified cluster.
-func (r *MigPlan) UpdateRegistryDC(storage *MigStorage, deploymentconfig *appsv1.DeploymentConfig, name, dirName string) error {
-	envVars := buildProxyEnvVars()
+func (r *MigPlan) UpdateRegistryDC(storage *MigStorage, deploymentconfig *appsv1.DeploymentConfig, proxySecret *kapi.Secret, name, dirName string) error {
+	envFrom := []kapi.EnvFromSource{}
+	// If Proxy secret exists, set env from it
+	if proxySecret != nil {
+		source := kapi.EnvFromSource{
+			SecretRef: &kapi.SecretEnvSource{
+				LocalObjectReference: kapi.LocalObjectReference{
```
Comment (jortel) [inline pkg/apis/migration/v1alpha1/migplan_types.go:430]: :+1:

File: pkg/apis/migration/v1alpha1/migplan_types.go (hunk 3 above)
Comment (dymurray) [reply pkg/apis/migration/v1alpha1/migplan_types.go:406]: I don't know :)

File: pkg/apis/migration/v1alpha1/migplan_types.go (hunk 4 above)
Comment (dymurray) [reply pkg/apis/migration/v1alpha1/migplan_types.go:403]: Done

File: pkg/controller/migplan/validation.go [hunk 6]
```diff
@@ -759,6 +771,115 @@ func (r ReconcileMigPlan) validatePvSelections(plan *migapi.MigPlan) error {
...
+	if plan.Status.HasAnyCondition(Suspended) {
+		return nil
+	}
+
+	list := kapi.SecretList{}
+	selector := k8sLabels.SelectorFromSet(map[string]string{
+		"migration-proxy-config": "true",
+	})
+
+	// Source cluster proxy secret validation
```
Comment (jortel) [inline pkg/controller/migplan/validation.go:786]: The comment suggests that `GetSourceCluster()` is doing validation?

File: pkg/controller/migplan/validation.go [hunk 7]
```diff
@@ -69,14 +72,15 @@ const (
...
-	NotDistinct   = "NotDistinct"
-	LimitExceeded = "LimitExceeded"
-	NotDone       = "NotDone"
-	Done          = "Done"
-	Conflict      = "Conflict"
-	NotHealthy    = "NotHealthy"
+	NotSet                     = "NotSet"
+	NotFound                   = "NotFound"
+	NotDistinct                = "NotDistinct"
+	ProxySecretIsMisconfigured = "ProxySecretIsMisconfigured"
```
Comment (jortel) [inline pkg/controller/migplan/validation.go:78]: The `reason` should describe the reason for the condition (or last state transition).

Perhaps:

`KeyError` when the secret is missing keys.
`Conflict` when multiple secrets found.

File: pkg/controller/migplan/validation.go [hunk 8]
```diff
@@ -759,6 +771,115 @@ func (r ReconcileMigPlan) validatePvSelections(plan *migapi.MigPlan) error {
...
+	misconfiguredClusters := 0
+	if plan.Status.HasAnyCondition(Suspended) {
+		return nil
+	}
+
+	list := kapi.SecretList{}
+	selector := k8sLabels.SelectorFromSet(map[string]string{
+		"migration-proxy-config": "true",
+	})
+
```
Comment (jortel) [inline pkg/controller/migplan/validation.go:785]: Following the pattern of other validations I anticipated:

Types:  _like_
- SourceClusterProxyMisconfigured
- DestinationClusterProxyMisconfigured

Reasons: _like_
- Conflict
- KeyNotFound

One method for each (source, destionation).

Each method could check for (using source cluster example):
```
if plan.Status.HasAnyCondition(Suspended, InvalidSourceClusterRef, SourceClusterNotReady) {
      return nil
}
...
if len(list.Items) > 1 {
    plan.Status.SetCondition(migapi.Condition{
        Type:     SourceClusterProxyMisconfigured,
        Status:   True,
        Reason:   Conflict,
        Category: Critical,
        Message:  SourceClusterProxyMisconfiguredMessage,
    })
    return nil
}
if !r.validateProxySecret(list.Items[0]) {
        plan.Status.SetCondition(migapi.Condition{
        Type:     SourceClusterProxyMisconfigured,
        Status:   True,
        Reason:   KeyNotFound,
        Category: Critical,
        Message:  SourceClusterProxyMisconfiguredMessage,
    })
}
```

```
func (r *ReconcileMigPlan) validateProxySecret(secret *kapi.Secret) bool {
    fields := []string{
        settings.HttpProxy,
        settings.HttpsProxy,
        settings.NoProxy,
    }
    for _, key := range fields {
        if _, found := secret.Data[key]; found {
            return true
        }
    }

    return false
}
```

File: pkg/controller/migplan/validation.go (hunk 7 above)
Comment (dymurray) [reply pkg/controller/migplan/validation.go:78]: Got it, thank you

File: pkg/controller/migplan/validation.go (hunk 8 above)
Comment (dymurray) [reply pkg/controller/migplan/validation.go:811]: Used a lot of this in the commit. Thank you very much for the thorough explanation.

File: pkg/controller/migplan/validation.go [hunk 9]
```diff
@@ -759,6 +773,157 @@ func (r ReconcileMigPlan) validatePvSelections(plan *migapi.MigPlan) error {
...
+	// Source cluster proxy secret validation
+	srcCluster, err := plan.GetSourceCluster(r)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	if srcCluster == nil || !srcCluster.Status.IsReady() {
```
Comment (jortel) [inline pkg/controller/migplan/validation.go:819]: The IsReady() here is fine but redundant with the HasAnyCondition() check on `SourceClusterNotReady`.

File: pkg/controller/migplan/validation.go [hunk 10]
```diff
@@ -759,6 +773,157 @@ func (r ReconcileMigPlan) validatePvSelections(plan *migapi.MigPlan) error {
...
+	})
+
+	// Destination cluster proxy secret validation
+	destCluster, err := plan.GetDestinationCluster(r)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	if destCluster == nil || !destCluster.Status.IsReady() {
```
Comment (jortel) [inline pkg/controller/migplan/validation.go:884]: The IsReady() here is fine but a little redundant with the HasAnyCondition() check on `DestinationClusterNotReady` above.

**Comment by dymurray** [comment]
Updated with Jeff's comments

**Pull request merged**